File: 24 Swap Nodes in Pairs/24.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        return reverseKGroup(head, 2)

def reverseKGroup(head, k):
        count = 0
        current = head
        while (current and count < k):
            current = current.next
            count += 1
        if (count == k):
            last = reverseKGroup(current, k)
            while (count > 0):
                next = head.next
                head.next = last
                # next.next = head <- don't need this line
                head, last = next, head
                count -= 1
            head = last
        return head

File: 24 Swap Nodes in Pairs/24.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        int count = 0;
        int k = 2;
        ListNode cur = head;
        while (cur != null && count < k) {
            cur = cur.next;
            count++;
        }
        if (count == k) {
            ListNode last = swapPairs(cur);
            ListNode nextNode = null;
            while (count-- > 0) {
                nextNode = head.next;
                head.next = last;
                last = head;
                head = nextNode;
            }
            head = last;
        }
        return head; 
    }
}